lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
+  for (auto p : llvm::enumerate(dti->getResults())) {
+    if (tupleElts[p.index()].get() != p.value())
+      return nextII;
+    if (!p.value()->hasOneUse())
+      return nextII;
+  }
+
+  // Then make everything that used the tuple, use the destructure's opeand and
+  // then delete the tuple/destructure in that order.
+  pass.callbacks.replaceValueUsesWith(tupleInst, dti->getOperand());
+  nextII = killInstruction(tupleInst, nextII, pass);
+  nextII = killInstruction(dti, nextII, pass);
+  return nextII;
+}
+
+//===----------------------------------------------------------------------===//
+//                        Destructure Simplifications
+//===----------------------------------------------------------------------===//
+
+// %x = tuple(hat(%x))
+// hat(%x2) = destructure_tuple x
+//
+// ->
+// hat(%x2) = hat(%x)
+static SILBasicBlock::iterator eliminateRoundTripDestructureOfTuple(
+    TupleInst *tupleInst, DestructureTupleInst *dti,
+    SILBasicBlock::iterator nextII, CanonicalizeInstruction &pass) {
+  // Make sure the types line up.
+  if (dti->getOperand()->getType() != tupleInst->getType())
+    return nextII;
+
+  // Make sure our tuple only has a single use, us.
+  if (!tupleInst->hasOneUse())
+    return nextII;
+
+  // Ok, we can eliminate this! Replace all uses of our dti's results with uses
+  // of the tupleInst's operands and then destroy the dti and the tuple in that
+  // order.
+  for (auto p : llvm::enumerate(dti->getAllResults())) {
+    p.value()->replaceAllUsesWith(tupleInst->getOperand(p.index()));
+  }
+  tupleInst->replaceAllUsesOfAllResultsWithUndef();
+  dti->replaceAllUsesOfAllResultsWithUndef();
+  nextII = killInstruction(dti, nextII, pass);
+  nextII = killInstruction(tupleInst, nextII, pass);
+  return nextII;
+}
+
+// We assume that we are going to see one of two patterns:
+//
+//   %alloc = alloc_stack $(Int, Int)
+//   ...
+//   %t0 = tuple_element_addr %alloc, 0
+//   %t1 = tuple_element_addr %alloc, 1
+//   (%0, %1) = destructure_tuple %x
+//   ...
+//   store %0 to %t0
+//   store %1 to %t1
+//
+// Or:
+//
+//   %alloc = alloc_stack $(Int, Int)
+//   ...
+//   (%0, %1) = destructure_tuple %x
+//   ...
+//   %t0 = tuple_element_addr %alloc, 0
+//   store %0 to %t0
+//   %t1 = tuple_element_addr %alloc, 1
+//   store %1 to %t1
+//
+// To:
+//
+//   %alloc = alloc_stack $(Int, Int)
+//   ...
+//   store %x to %alloc
+//
+// We assume everything lines up and there aren't any intervening instructions
+// since this is just looking for code out of SILGen.
+//
+// We iterate over the code sequence twice, once to check for properties and
+// then to optimize instead of using a seenInsts or the like.
+static SILBasicBlock::iterator
+canonicalizeAwayTupleEltAddrDestructurePairs(DestructureTupleInst *dti,
+                                             SILBasicBlock::iterator next,
+                                             CanonicalizeInstruction &pass) {
+  if (dti->getModule().getStage() != SILStage::Raw)
```

Comment:
Is it really required to be in raw stage? It would probably not harm to try this transformation also on canonical SIL. On the other hand, it's really a very specific pattern matching, which is unlikely to be generated by an optimization - I guess.
Anyway, can you add a comment why this is done only on raw SIL?

Reply:
The reason why I did this only in Raw SIL is exactly due to what you said. I can add a comment explaining that it is b/c it is so specific to the SILGen output that it probably won't be useful otherwise. I was just trying to save compile time by early exiting.